### PR TITLE
Feature sns integration

### PR DESCRIPTION
**Pull request by subodhranadive**
**added**
- Amazon SNS SMS support

**Documentation**

Configuration added

- `aws.sns.enabled` : Use this key to send message using SNS. Accepted value - enable/disable
- `aws.access` : AWS Access Key with SNS permission
- `aws.secret` : AWS Secret Access Key with SNS permission
- `aws.region` : AWS Region for SNS service. **NOTE** : Make sure to use regions that are supported for SNS messaging.

To enable sns messaging, we need to make sure that the value `sms.enable` is `true` and `notificator.types` contains `sms`.

### CODE REVIEW

**Comment by subodhranadive** [comment]
@tananaev Please can you take a look at this PR.

Let me know if there are any changes required from my end.

Thanks,
Subodh

File: build.gradle [hunk 1]
```diff
@@ -77,6 +77,7 @@ dependencies {
     implementation "com.sun.xml.bind:jaxb-core:2.3.0.1"
     implementation "com.sun.xml.bind:jaxb-impl:2.3.3"
     implementation "javax.activation:activation:1.1.1"
+    implementation 'com.amazonaws:aws-java-sdk-sns:1.11.692'
```
Comment (tananaev) [inline build.gradle:80]: Is there an API we can use instead of integrating a library?

Comment (subodhranadive) [reply build.gradle:80]: Thanks for responding @tananaev.
It would be better if we use SDK instead of using bare API itself. As we underlying authentication/authorisation and other complexities are taken care by the SDK itself. Also, we wont have to deal with the security aspect as well in case of SDK.

Comment (tananaev) [reply build.gradle:80]: Is the SDK open source?

Comment (subodhranadive) [reply build.gradle:80]: Yes the SDK is open source.

**Ref**: [AWS Java SDK](https://github.com/aws/aws-sdk-java)

File: src/main/java/org/traccar/config/Keys.java [hunk 2]
```diff
@@ -687,6 +687,36 @@ private Keys() {
             "sms.http.template",
             Collections.singletonList(KeyType.GLOBAL));
 
+    /**
+     * AWS Access Key with SNS permission.
+     */
+    public static final ConfigKey<String> AWS_ACCESS_KEY = new ConfigKey<>(
+            "aws.access",
```
Comment (tananaev) [inline src/main/java/org/traccar/config/Keys.java:694]: Let's name all parameters with a prefix `sms.aws.`.

File: src/main/java/org/traccar/config/Keys.java [hunk 3]
```diff
@@ -687,6 +687,36 @@ private Keys() {
...
+    public static final ConfigKey<String> AWS_SECRET_KEY = new ConfigKey<>(
+            "aws.secret",
+            Collections.singletonList(KeyType.GLOBAL));
+
+    /**
+     * SNS Enable switch.
+     * Use 'enable' to send message with SNS or 'disable' to stop using SNS
+     */
+    public static final ConfigKey<String> AWS_SNS_ENABLED = new ConfigKey<>(
+            "aws.sns.enabled",
```
Comment (tananaev) [inline src/main/java/org/traccar/config/Keys.java:709]: We are trying to reduce number of config parameters. Let's use presence of one of the other keys as "enabled" flag.

File: src/main/java/org/traccar/sms/SnsSmsClient.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
+package org.traccar.sms;
```
Comment (tananaev) [inline src/main/java/org/traccar/sms/SnsSmsClient.java:1]: Please add license notice here.

File: src/main/java/org/traccar/sms/SnsSmsClient.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        snsStatus = Context.getConfig().getString(Keys.AWS_SNS_ENABLED);
+        region = Context.getConfig().getString(Keys.AWS_REGION);
+        snsClient = awsSNSClient(accessKey, secretKey, region);
+
+        if (!snsStatus.equals("true") || accessKey == null || secretKey == null || region == null) {
+            LOGGER.error("SNS Not Configured Properly. Please provide valid config.");
+        }
+    }
+
+    public AmazonSNS awsSNSClient(String accessKey, String secretKey, String region) {
```
Comment (tananaev) [inline src/main/java/org/traccar/sms/SnsSmsClient.java:42]: This can probably be inlined and we won't need to keep config parameters as class variables.

File: src/main/java/org/traccar/sms/SnsSmsClient.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        BasicAWSCredentials awsCreds = new BasicAWSCredentials(accessKey, secretKey);
+        return AmazonSNSClientBuilder.standard().withRegion(region)
+                .withCredentials(new AWSStaticCredentialsProvider(awsCreds)).build();
+    }
+
+    public void sendSNSMessage(String message, String destAddress) {
```
Comment (tananaev) [inline src/main/java/org/traccar/sms/SnsSmsClient.java:48]: Why do we need this method? Can't we just put this code in `sendMessageSync`?

File: src/main/java/org/traccar/sms/SnsSmsClient.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+        Map<String, MessageAttributeValue> smsAttributes = new HashMap<>();
+        smsAttributes.put("AWS.SNS.SMS.SenderID",
+                new MessageAttributeValue().withStringValue("SNS").withDataType("String"));
+        smsAttributes.put("AWS.SNS.SMS.SMSType",
+                new MessageAttributeValue().withStringValue("Transactional").withDataType("String"));
+
+        PublishResult result = this.snsClient.publish(new PublishRequest().withMessage(message)
```
Comment (tananaev) [inline src/main/java/org/traccar/sms/SnsSmsClient.java:55]: Remove `this`.

File: src/main/java/org/traccar/sms/SnsSmsClient.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+                .withPhoneNumber(destAddress).withMessageAttributes(smsAttributes));
+    }
+
+    @java.lang.Override
```
Comment (tananaev) [inline src/main/java/org/traccar/sms/SnsSmsClient.java:59]: Why do you need `java.lang.`?

File: src/main/java/org/traccar/config/Keys.java (hunk 3 above)
Comment (subodhranadive) [reply src/main/java/org/traccar/config/Keys.java]: I have gone through the config variables but I couldn't find one which represents the specific/meaningful description to enable/disable this feature.

File: src/main/java/org/traccar/sms/SnsSmsClient.java (hunk 6 above)
Comment (subodhranadive) [reply src/main/java/org/traccar/sms/SnsSmsClient.java:57]: To minimize the amount of code duplication I have implemented this method, as we are using the same code in `sendMessageSync` and `sendMessageAsync`.

Comment (tananaev) [reply src/main/java/org/traccar/sms/SnsSmsClient.java:56]: You can just put it directly in the `sendMessageSync`. And call that method from `sendMessageAsync`. I think we do it in some other places.

File: src/main/java/org/traccar/config/Keys.java (hunk 3 above)
Comment (tananaev) [reply src/main/java/org/traccar/config/Keys.java]: I meant to use other keys, like `region` or something.

**Comment by subodhranadive** [comment]
@tananaev I have made changes as requested. Can you please review it and let me know if there are any changes required.

Thanks,
Subodh

File: src/main/java/org/traccar/config/Keys.java [hunk 9]
```diff
@@ -687,6 +687,28 @@ private Keys() {
             "sms.http.template",
             Collections.singletonList(KeyType.GLOBAL));
 
+    /**
+     * AWS Access Key with SNS permission.
+     */
+    public static final ConfigKey<String> AWS_ACCESS_KEY = new ConfigKey<>(
```
Comment (tananaev) [inline src/main/java/org/traccar/config/Keys.java:693]: Please make sure variable names match keys. For example this should be `SMS_AWS_ACCESS`.

File: src/main/java/org/traccar/protocol/HuabaoProtocolDecoder.java [hunk 10]
```diff
@@ -171,7 +172,7 @@ protected Object decode(
                         formatMessage(MSG_TERMINAL_REGISTER_RESPONSE, id, false, response), remoteAddress));
             }
 
-        } else if (type == MSG_TERMINAL_AUTH) {
+        } else if (type == MSG_TERMINAL_AUTH || type == MSG_HEARTBEAT) {
```
Comment (tananaev) [inline src/main/java/org/traccar/protocol/HuabaoProtocolDecoder.java:175]: What is this change?

File: src/main/java/org/traccar/protocol/Gt06ProtocolDecoder.java [hunk 11]
```diff
@@ -838,7 +838,9 @@ private void decodeHeartbeat(ByteBuf buf, Position position) {
         buf.readUnsignedShort(); // satellites
         buf.readUnsignedByte(); // alarm
         buf.readUnsignedByte(); // language
-        buf.readUnsignedByte(); // battery
+
+        position.set(Position.KEY_BATTERY_LEVEL, buf.readUnsignedByte());
```
Comment (tananaev) [inline src/main/java/org/traccar/protocol/Gt06ProtocolDecoder.java:842]: And this?

File: src/main/java/org/traccar/sms/SnsSmsClient.java [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+import java.util.Map;
+
+
+public class SnsSmsClient implements SmsManager {
+    private static final Logger LOGGER = LoggerFactory.getLogger(SnsSmsClient.class);
+
+    private final AmazonSNS snsClient;
+
+    public SnsSmsClient() {
+        if (Context.getConfig().getString(Keys.AWS_REGION) == null
```
Comment (tananaev) [inline src/main/java/org/traccar/sms/SnsSmsClient.java:40]: There's `hasKey` that you should be using instead.

File: src/main/java/org/traccar/sms/SnsSmsClient.java [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+                || Context.getConfig().getString(Keys.AWS_ACCESS_KEY) == null
+                || Context.getConfig().getString(Keys.AWS_SECRET_KEY) == null) {
+            throw new RuntimeException("SNS Not Configured Properly. Please provide valid config.");
+        }
+        snsClient = awsSNSClient();
+    }
+
+    public AmazonSNS awsSNSClient() {
```
Comment (tananaev) [inline src/main/java/org/traccar/sms/SnsSmsClient.java:48]: What's the reason for this method?

File: src/main/java/org/traccar/sms/SnsSmsClient.java [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+                .withPhoneNumber(destAddress).withMessageAttributes(smsAttributes));
+    }
+
+    @Override
+    public void sendMessageAsync(String destAddress, String message, boolean command) {
+        try {
+            sendMessageSync(destAddress, message, command);
+        } catch (InterruptedException interruptedException) {
+            LOGGER.warn("SMS send failed", interruptedException.getMessage());
+        } catch (MessageException messageException) {
```
Comment (tananaev) [inline src/main/java/org/traccar/sms/SnsSmsClient.java:73]: Why do you need two catch blocks that do the same thing?

File: src/test/java/org/traccar/protocol/Gt06ProtocolDecoderTest.java [hunk 15]
```diff
@@ -17,6 +17,10 @@ public void testDecode() throws Exception {
         verifyNull(decoder, binary(
                 "78780D01086471700328358100093F040D0A"));
 
+        verifyAttribute(decoder, binary(
```
Comment (tananaev) [inline src/test/java/org/traccar/protocol/Gt06ProtocolDecoderTest.java:20]: This also doesn't seem to be related.

**Review by tananaev — COMMENTED** [review]
Added some comments. Looks like you have some unrelated changes included.

File: src/main/java/org/traccar/sms/SnsSmsClient.java (hunk 13 above)
Comment (subodhranadive) [reply src/main/java/org/traccar/sms/SnsSmsClient.java:49]: To abstract the AWS SNS client so that we can use just the same client in future if we need to add new feature and we can use the same method.

**Comment by subodhranadive** [comment]
@tananaev I have updated the PR. Please can you review it and let me know if there are any changes required.

Thanks,
Subodh

File: src/main/java/org/traccar/sms/SnsSmsClient.java (hunk 13 above)
Comment (tananaev) [reply src/main/java/org/traccar/sms/SnsSmsClient.java:49]: This part doesn't really make sense to me. Please inline for now.

File: src/main/java/org/traccar/sms/SnsSmsClient.java [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+
+    @Override
+    public void sendMessageSync(String destAddress, String message, boolean command)
+            throws InterruptedException, MessageException {
+        Map<String, MessageAttributeValue> smsAttributes = new HashMap<>();
+        smsAttributes.put("AWS.SNS.SMS.SenderID",
+                new MessageAttributeValue().withStringValue("SNS").withDataType("String"));
+        smsAttributes.put("AWS.SNS.SMS.SMSType",
+                new MessageAttributeValue().withStringValue("Transactional").withDataType("String"));
+        snsClient.publish(new PublishRequest().withMessage(message)
```
Comment (tananaev) [inline src/main/java/org/traccar/sms/SnsSmsClient.java:64]: Wait. Is this method asynchronous or not?

Comment (subodhranadive) [reply src/main/java/org/traccar/sms/SnsSmsClient.java:64]: Yes, this method is `asynchronous`. While publishing the message AWS returns request id and behind the scenes use `asynchronous` processing for sending the actual message. (as far as i know, AWS is not providing any `synchronous` method for sending sns messages)

Comment (tananaev) [reply src/main/java/org/traccar/sms/SnsSmsClient.java:64]: OK, can you please provide some reference to confirm that it's `asynchronous`. Also, I'm not sure how you asynchronous method can return those exceptions.

Comment (subodhranadive) [reply src/main/java/org/traccar/sms/SnsSmsClient.java:64]: **Reference** : [AWS SNS Publish Message](https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/sns/AmazonSNSClient.html#publish-com.amazonaws.services.sns.model.PublishRequest-)

As described in the documentation, it just returns messageId and processes the request.

`When a messageId is returned, the message has been saved and Amazon SNS will attempt to deliver it shortly.`

Comment (tananaev) [reply src/main/java/org/traccar/sms/SnsSmsClient.java:64]: Just want to clarify. I'm talking about network request to Amazon SNS. I'm not talking about flow on the Amazon server. It sounds like the documentation you are referencing is referring to the latter.

Comment (subodhranadive) [reply src/main/java/org/traccar/sms/SnsSmsClient.java:64]: When it comes to network call, it is synchronous.

What we can do is, we can create another method and make it as Async and call `sendMessageSync(...)` method from the newly created method. - For this we will have to use `@Async` annotation which requires `spring-context` dependency.

Please share your thoughts on the same.

Comment (tananaev) [reply src/main/java/org/traccar/sms/SnsSmsClient.java:64]: I don't think we need to include Sprint to just make a single asynchronous method. I think just starting a new thread would be enough.

Comment (subodhranadive) [reply src/main/java/org/traccar/sms/SnsSmsClient.java:62]: I have updated PR with required changes.

Thanks,
Subodh

**Comment by tananaev** [comment]
Looks good. We'll merge it before the next release.

**Comment by subodhranadive** [comment]
> Looks good. We'll merge it before the next release.

Thanks @tananaev

**Comment by subodhranadive** [comment]
**Documentation Update**

Configuration added

- `aws.region` : AWS Region for SNS service.
- `aws.access` : AWS Access Key with SNS permission
- `aws.secret` : AWS Secret Access Key with SNS permission

To enable sns messaging, we need to make sure that the value `sms.enable` is `true` and `notificator.types` contains `sms`.

**NOTE** : 
- Make sure to use regions that are supported for SNS messaging in `aws.region`.
- `aws.region` : This key will enable/disable SNS messaging. (If this key is present, then only SNS messaging will work).

**Comment by jcardus** [comment]
Just for the record, there's an async class
<https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/sns/AmazonSNSAsyncClient.html>
for that.

**Comment by tananaev** [comment]
Great, then we have to use that instead. @subodhranadive can you please update the PR.

**Comment by subodhranadive** [comment]
> Great, then we have to use that instead. @subodhranadive can you please update the PR.

@tananaev I have updated the PR. Please can you review it.

**Comment by jcardus** [comment]
hi @subodhranadive, looks like you're ignoring the result and possibly exceptions

File: src/main/java/org/traccar/sms/SnsSmsClient.java [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+        if (Context.getConfig().hasKey(Keys.SMS_AWS_REGION)
+                && Context.getConfig().hasKey(Keys.SMS_AWS_ACCESS)
+                && Context.getConfig().hasKey(Keys.SMS_AWS_SECRET)) {
+            snsClient = awsSNSClient();
+        } else {
+            throw new RuntimeException("SNS Not Configured Properly. Please provide valid config.");
+        }
+    }
+
+    public AmazonSNSAsync awsSNSClient() {
```
Comment (tananaev) [inline src/main/java/org/traccar/sms/SnsSmsClient.java:48]: Please inline this.

File: src/main/java/org/traccar/sms/SnsSmsClient.java [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    @Override
+    public void sendMessageSync(String destAddress, String message, boolean command) {
+        Map<String, MessageAttributeValue> smsAttributes = new HashMap<>();
+        smsAttributes.put("AWS.SNS.SMS.SenderID",
+                new MessageAttributeValue().withStringValue("SNS").withDataType("String"));
+        smsAttributes.put("AWS.SNS.SMS.SMSType",
+                new MessageAttributeValue().withStringValue("Transactional").withDataType("String"));
+        snsClient.publishAsync(new PublishRequest().withMessage(message)
```
Comment (tananaev) [inline src/main/java/org/traccar/sms/SnsSmsClient.java:62]: If there's an error, we should log something.

File: src/main/java/org/traccar/sms/SnsSmsClient.java [hunk 19]
```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2021 Anton Tananaev (anton@traccar.org)
```
Comment (tananaev) [inline src/main/java/org/traccar/sms/SnsSmsClient.java:2]: Feel free to put your name/email here if you want.

File: src/main/java/org/traccar/sms/SnsSmsClient.java (hunk 17 above)
Comment (subodhranadive) [reply src/main/java/org/traccar/sms/SnsSmsClient.java:47]: Can you please elaborate a little more. Perhaps an example.

Comment (tananaev) [reply src/main/java/org/traccar/sms/SnsSmsClient.java:49]: Remove this method. I already had a comment about it, but it got outdated, but not addressed.

**Comment by subodhranadive** [comment]
@tananaev I have updated the PR with suggested changes.

File: src/main/java/org/traccar/sms/SnsSmsClient.java [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+    public void sendMessageSync(String destAddress, String message, boolean command) {
+        Map<String, MessageAttributeValue> smsAttributes = new HashMap<>();
+        smsAttributes.put("AWS.SNS.SMS.SenderID",
+                new MessageAttributeValue().withStringValue("SNS").withDataType("String"));
+        smsAttributes.put("AWS.SNS.SMS.SMSType",
+                new MessageAttributeValue().withStringValue("Transactional").withDataType("String"));
+        try {
+            snsClient.publishAsync(new PublishRequest().withMessage(message)
+                    .withPhoneNumber(destAddress).withMessageAttributes(smsAttributes));
+        } catch (Exception exception) {
```
Comment (tananaev) [inline src/main/java/org/traccar/sms/SnsSmsClient.java:64]: I don't think that's how you handle asynchronous errors. This seems incorrect. Also, why is it a generic `Exception`?

**Review by tananaev — COMMENTED** [review]
Found an issues with error handling.

File: src/main/java/org/traccar/sms/SnsSmsClient.java (hunk 20 above)
Comment (subodhranadive) [reply src/main/java/org/traccar/sms/SnsSmsClient.java]: Hi @tananaev 
I have changed the generic `Exception` to `RuntimeException`.

Regarding handling of asynchronous errors, I took the reference of `publishAsync()` method of `AmazonSNSAsyncClient.java` as below 

[https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-sns/src/main/java/com/amazonaws/services/sns/AmazonSNSAsyncClient.java#L1337-L1362](https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-sns/src/main/java/com/amazonaws/services/sns/AmazonSNSAsyncClient.java#L1337-L1362)

Here they are throwing the exception and as we have added catch block in our code and as a result that will be handled.

Comment (tananaev) [reply src/main/java/org/traccar/sms/SnsSmsClient.java]: But it is thrown in a function that is presumably executed asynchronously. You cannot catch that here.

Comment (subodhranadive) [reply src/main/java/org/traccar/sms/SnsSmsClient.java]: Hi @tananaev 

Can we use something like below:

```
Thread.setDefaultUncaughtExceptionHandler((thread, exception) ->
                LOGGER.error("SMS send failed", exception));

        new Thread(() -> {
            Map<String, MessageAttributeValue> smsAttributes = new HashMap<>();
            smsAttributes.put("AWS.SNS.SMS.SenderID",
                    new MessageAttributeValue().withStringValue("SNS").withDataType("String"));
            smsAttributes.put("AWS.SNS.SMS.SMSType",
                    new MessageAttributeValue().withStringValue("Transactional").withDataType("String"));
            snsClient.publishAsync(new PublishRequest().withMessage(message)
                    .withPhoneNumber(destAddress).withMessageAttributes(smsAttributes));
        });
```
If not, Please can you provide a small example/snippet.

Comment (tananaev) [reply src/main/java/org/traccar/sms/SnsSmsClient.java]: I would think that because it's already `publishAsync`, you don't need a new thread. I don't have a snippet obviously.

Comment (subodhranadive) [reply src/main/java/org/traccar/sms/SnsSmsClient.java]: Hi @tananaev 

I have added AWS `AsyncHandler` for exception handling of `publishAsync()` method.

File: src/main/java/org/traccar/sms/SnsSmsClient.java [hunk 21]
```diff
@@ -0,0 +1,83 @@
...
+                .withPhoneNumber(destAddress).withMessageAttributes(smsAttributes);
+
+        snsClient.publishAsync(publishRequest, new AsyncHandler<PublishRequest, PublishResult>() {
+            @Override
+            public void onError(Exception exception) {
+                LOGGER.error("SMS send failed", exception);
+            }
+            @Override
+            public void onSuccess(PublishRequest request, PublishResult result) {
+                LOGGER.info("SMS sent successfully");
```
Comment (tananaev) [inline src/main/java/org/traccar/sms/SnsSmsClient.java:74]: Please don't log anything on success.

**Review by tananaev — COMMENTED** [review]
Looks much better. Just one comment.

File: src/main/java/org/traccar/sms/SnsSmsClient.java (hunk 21 above)
Comment (subodhranadive) [reply src/main/java/org/traccar/sms/SnsSmsClient.java]: Hi @tananaev 

I have updated the PR.

**Comment by tananaev** [comment]
Merged, thank you.

**Comment by subodhranadive** [comment]
> Merged, thank you.

Thanks @tananaev
